module/executiondatasync/optimistic_sync/pipeline_functional_test.go

```diff
@@ -0,0 +1,447 @@
+	})
+}
+
+// TestMainCtxCancellationDuringRequestingTxResultErrMsgs tests context cancellation during
+// the request of transaction result error messages. It verifies that when the parent context
+// is cancelled during this phase, the pipeline handles the cancellation gracefully
+// and transitions to the correct state.
+func (p *PipelineFunctionalSuite) TestMainCtxCancellationDuringRequestingTxResultErrMsgs() {
+	p.WithRunningPipeline(func(pipeline Pipeline, updateChan chan State, errChan chan error, cancel context.CancelFunc) {
+		pipeline.OnParentStateUpdated(StateComplete)
+
+		p.execDataRequester.On("RequestExecutionData", mock.Anything).Return((*execution_data.BlockExecutionData)(nil), nil).Maybe()
+
+		//This call marked as `Maybe`m because it may not be called depending on timing.
+		p.txResultErrMsgsRequester.On("Request", mock.Anything).Return(
+			func(ctx context.Context) ([]flow.TransactionResultErrorMessage, error) {
+				// Wait for cancellation
+				cancel()
+
+				<-ctx.Done()
+
+				return nil, ctx.Err()
+			}).Once()
+
+		waitForStateUpdates(p.T(), updateChan, StateProcessing)
+		waitForError(p.T(), errChan, context.Canceled)
+	})
+}
+
+// TestPipelineShutdownOnParentAbandon tests pipeline behavior when the parent state
+// is abandoned. It verifies that the pipeline properly shuts down, cancels ongoing
+// operations, and cleans up its working data when the parent abandons the operation.
+func (p *PipelineFunctionalSuite) TestPipelineShutdownOnParentAbandon() {
+	expectedExecutionData, expectedTxResultErrMsgs := p.initializeTestData()
+
+	p.execDataRequester.On("RequestExecutionData", mock.Anything).Return(expectedExecutionData, nil).Once()
+	p.txResultErrMsgsRequester.On("Request", mock.Anything).Return(expectedTxResultErrMsgs, nil).Once()
+
+	p.WithRunningPipeline(func(pipeline Pipeline, updateChan chan State, errChan chan error, cancel context.CancelFunc) {
+		waitForStateUpdates(p.T(), updateChan, StateProcessing, StateWaitingPersist)
+
+		pipeline.OnParentStateUpdated(StateAbandoned)
+
+		waitForStateUpdates(p.T(), updateChan, StateAbandoned)
+		waitForError(p.T(), errChan, nil)
+
+		p.Assert().Equal(StateAbandoned, pipeline.GetState())
+		p.Assert().Nil(p.core.workingData)
+	})
+}
+
+// WithRunningPipeline is a test helper that initializes and starts a pipeline instance.
+// It manages the context and channels needed to run the pipeline and invokes the testFunc
+// with access to the pipeline, update channel, error channel, and cancel function.
+func (p *PipelineFunctionalSuite) WithRunningPipeline(testFunc func(pipeline Pipeline, updateChan chan State, errChan chan error, cancel context.CancelFunc)) {
+	pipelineStateConsumer := NewMockStateConsumer()
+
+	pipeline := NewPipeline(p.logger, p.executionResult, false, pipelineStateConsumer)
+
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	errChan := make(chan error)
+	//wait until a pipeline goroutine run a pipeline
```

Comment:
always include a space before the text in comments (except in //nolint and //goembed directives)
```suggestion
	// wait until a pipeline goroutine run a pipeline
```